collect_app/src/test/java/org/odk/collect/android/widgets/UrlWidgetTest.java

```diff
@@ -1,46 +1,146 @@
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptAnswerDoesNotHaveAnswer_displayEmptyString() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo(""));
+    }
+
+    @Test
+    public void clickingButtonWhenUrlIsEmpty_doesNotCallOpenUri() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper, never()).bindCustomTabsService(null, null);
+        verify(customTabHelper, never()).openUri(null, null);
+    }
+
+    @Test
+    public void clickingButton_callsCorrectMethods() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper).bindCustomTabsService(widget.getContext(), null);
+        verify(customTabHelper).openUri(widget.getContext(), Uri.parse("blah"));
     }
 
-    @NonNull
-    @Override
-    public StringData getNextAnswer() {
-        return new StringData(RandomString.make());
+    @Test
+    public void clickingButtonForLong_callsLongClickListener() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        widget.setOnLongClickListener(listener);
+        urlButton.performLongClick();
+        verify(listener).onLongClick(urlButton);
     }
 
-    @Override
-    public void callingClearShouldRemoveTheExistingAnswer() {
-        // The widget is ReadOnly, clear shouldn't do anything.
+    @Test
+    public void cancelLongPress_callsCancelLongPressForButtonAndTextView() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        widget.openUrlButton = mock(Button.class);
+        widget.stringAnswer = mock(TextView.class);
+        widget.cancelLongPress();
+
+        verify(widget.openUrlButton).cancelLongPress();
+        verify(widget.stringAnswer).cancelLongPress();
     }
 
     @Test
-    public void usingReadOnlyOptionShouldMakeAllClickableElementsDisabled() {
-        when(formEntryPrompt.isReadOnly()).thenReturn(true);
+    public void detachingFromWindow_disconnectsService_whenServiceConnectionIsNotNull() {
+        when(customTabHelper.getServiceConnection()).thenReturn(serviceConnection);
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        widget.onDetachedFromWindow();
+        verify(serviceConnection).onServiceDisconnected(null);
+    }
+
+    @Test
+    public void detachingFromWindow_doesNotCallOnServiceDisconnected_whenServiceConnectionIstNull() {
```

Comment:
Extra 't' in `whenServiceConnectionIstNull`

Reply:
Ahh.. sorry for the typo.